### ISSUE TITLE
微调bge-m3

### ISSUE THREAD

**Issue by WangCC-77** [FlagOpen/FlagEmbedding#1195]
作者，您好！非常感谢你们能够开源这么棒的模型，我现阶段想要复现bge-m3的微调，目前手上有的资源是8张 V100（32G）。具体的执行语句是：
`torchrun --nproc_per_node 8 \
-m FlagEmbedding.BGE_M3.run \
--output_dir ../../output/finetune/firstModel \
--model_name_or_path ./preModel/bge-m3-unsupervised/ \
--train_data  ../../data/bge-m3-data-test/  \
--learning_rate 1e-5 \
--fp16 \
--num_train_epochs 5 \
--per_device_train_batch_size 1 \
--dataloader_drop_last True \
--normlized True \
--temperature 0.02 \
--query_max_len 512 \
--passage_max_len 8192 \
--train_group_size 8 \
--negatives_cross_device \
--logging_steps 10 \
--same_task_within_batch True \
--unified_finetuning True \
--use_self_distill True `
使用的基座模型是**bge-m3-unsupervised**，文中说到使用了24 张A800（80G）进行训练，我按照相应的比例对batch_size进行了缩放，
train_group_size = 8，缩放比例是0.4：
![image](https://github.com/user-attachments/assets/326d9ea1-d1ac-4274-b706-842dc56d5681)
缩放比例是0.3：
![image](https://github.com/user-attachments/assets/a5a31eec-a2b1-47f9-bc95-05534fd08510)
缩放比例是0.1：
![image](https://github.com/user-attachments/assets/5acca90f-10bb-44b7-a8c2-7ca67a059fb1)
以上缩放比例均会报OOM的错误，具体报错代码位置：

目前采取的措施是根据显存容量对batch_size进行缩放，但还是跑不通。想要问问作者关于这个问题还有什么解决措施，或者说我还有哪里需要去改进，比如模型运行过程中是否涉及工程问题？麻烦作者帮忙看看，感谢。

**Comment by WangCC-77** [comment]
![20241106-140411(WeLinkPC)](https://github.com/user-attachments/assets/fd85ba66-2e80-4e5e-87e7-86f1ddec55b9)

**Comment by 545999961** [comment]
可以尝试使用gradient_checkpoint和deepspeed来节省更多的显存

**Comment by WangCC-77** [comment]
@545999961 谢谢你的解答，我想问问使用gradient_checkpoint具体在哪个地方呀，我搜了一下有关于TrainingArguments里面的gradient_checkpointing设置为True，还有一个是PyTorch的gradient checkpoint是通过torch.utils.checkpoint.checkpoint(function, *args, **kwargs)函数实现的。不知道该使用哪一个，麻烦再帮忙看看呀，谢谢。

**Comment by 545999961** [comment]
> @545999961 谢谢你的解答，我想问问使用gradient_checkpoint具体在哪个地方呀，我搜了一下有关于TrainingArguments里面的gradient_checkpointing设置为True，还有一个是PyTorch的gradient checkpoint是通过torch.utils.checkpoint.checkpoint(function, *args, **kwargs)函数实现的。不知道该使用哪一个，麻烦再帮忙看看呀，谢谢。

使用gradient_checkpoint True参数就可以，可以参考[微调](https://github.com/FlagOpen/FlagEmbedding/tree/master/examples#1-embedder-1)

**Comment by WangCC-77** [comment]
> > @545999961 谢谢你的解答，我想问问使用gradient_checkpoint具体在哪个地方呀，我搜了一下有关于TrainingArguments里面的gradient_checkpointing设置为True，还有一个是PyTorch的gradient checkpoint是通过torch.utils.checkpoint.checkpoint(function, *args, **kwargs)函数实现的。不知道该使用哪一个，麻烦再帮忙看看呀，谢谢。
> 
> 使用gradient_checkpoint True参数就可以，可以参考[微调](https://github.com/FlagOpen/FlagEmbedding/tree/master/examples#1-embedder-1)

十分感谢你的答复，今天和导师讨论了一下，他说我这边在进行模型加载的时候就爆内存了，还没到梯度训练那一块儿。说是这个报错可能和我目前环境安装的transformers版本（4.20.0）和你们给的（>=4.33.0）不一致，不知道是不是这个原因呢，打算重新换个环境再试试。

**Comment by WangCC-77** [comment]
> > @545999961 谢谢你的解答，我想问问使用gradient_checkpoint具体在哪个地方呀，我搜了一下有关于TrainingArguments里面的gradient_checkpointing设置为True，还有一个是PyTorch的gradient checkpoint是通过torch.utils.checkpoint.checkpoint(function, *args, **kwargs)函数实现的。不知道该使用哪一个，麻烦再帮忙看看呀，谢谢。
> 
> 使用gradient_checkpoint True参数就可以，可以参考[微调](https://github.com/FlagOpen/FlagEmbedding/tree/master/examples#1-embedder-1)

@545999961作者，你好。我这边按照你的指示加了gradient_checkpoint和deepspeed，我的环境是：
torch  2.4.0+cu118；torchaudio 2.4.0+cu118；torchvision    0.19.0+cu118；deepspeed  0.15.3；python 3.9
目前的报错是：
![image](https://github.com/user-attachments/assets/21b54607-e8d9-429d-a536-8286063ccede)
麻烦作者帮忙看看呢，谢谢~

**Comment by 545999961** [comment]
> > > @545999961 谢谢你的解答，我想问问使用gradient_checkpoint具体在哪个地方呀，我搜了一下有关于TrainingArguments里面的gradient_checkpointing设置为True，还有一个是PyTorch的gradient checkpoint是通过torch.utils.checkpoint.checkpoint(function, *args, **kwargs)函数实现的。不知道该使用哪一个，麻烦再帮忙看看呀，谢谢。
> > 
> > 
> > 使用gradient_checkpoint True参数就可以，可以参考[微调](https://github.com/FlagOpen/FlagEmbedding/tree/master/examples#1-embedder-1)
> 
> @545999961作者，你好。我这边按照你的指示加了gradient_checkpoint和deepspeed，我的环境是： torch 2.4.0+cu118；torchaudio 2.4.0+cu118；torchvision 0.19.0+cu118；deepspeed 0.15.3；python 3.9 目前的报错是： ![image](https://private-user-images.githubusercontent.com/75521911/384214611-21b54607-e8d9-429d-a536-8286063ccede.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwNTE2OTgsIm5iZiI6MTczMTA1MTM5OCwicGF0aCI6Ii83NTUyMTkxMS8zODQyMTQ2MTEtMjFiNTQ2MDctZThkOS00MjlkLWE1MzYtODI4NjA2M2NjZWRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA4VDA3MzYzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJhN2EwZjNmZTk5YjRhMDhmNDVlODFkMTdhYmQxY2M3ZGFlMzgyZDM2MjNlMTMwNDhlNDM1NTNlNjhmZmJkYzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yLmR6aJMoRjLGtRAFosHffgJOTXVXIPEdyY4G9UNfis) 麻烦作者帮忙看看呢，谢谢~

看起来像是deepspeed的问题，可以尝试更新deepspeed或更换使用的deepspeed配置文件

**Comment by WangCC-77** [comment]
> > > > @545999961 谢谢你的解答，我想问问使用gradient_checkpoint具体在哪个地方呀，我搜了一下有关于TrainingArguments里面的gradient_checkpointing设置为True，还有一个是PyTorch的gradient checkpoint是通过torch.utils.checkpoint.checkpoint(function, *args, **kwargs)函数实现的。不知道该使用哪一个，麻烦再帮忙看看呀，谢谢。
> > > 
> > > 
> > > 使用gradient_checkpoint True参数就可以，可以参考[微调](https://github.com/FlagOpen/FlagEmbedding/tree/master/examples#1-embedder-1)
> > 
> > 
> > @545999961作者，你好。我这边按照你的指示加了gradient_checkpoint和deepspeed，我的环境是： torch 2.4.0+cu118；torchaudio 2.4.0+cu118；torchvision 0.19.0+cu118；deepspeed 0.15.3；python 3.9 目前的报错是： ![image](https://private-user-images.githubusercontent.com/75521911/384214611-21b54607-e8d9-429d-a536-8286063ccede.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwNTE2OTgsIm5iZiI6MTczMTA1MTM5OCwicGF0aCI6Ii83NTUyMTkxMS8zODQyMTQ2MTEtMjFiNTQ2MDctZThkOS00MjlkLWE1MzYtODI4NjA2M2NjZWRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA4VDA3MzYzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJhN2EwZjNmZTk5YjRhMDhmNDVlODFkMTdhYmQxY2M3ZGFlMzgyZDM2MjNlMTMwNDhlNDM1NTNlNjhmZmJkYzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yLmR6aJMoRjLGtRAFosHffgJOTXVXIPEdyY4G9UNfis) 麻烦作者帮忙看看呢，谢谢~
> 
> 看起来像是deepspeed的问题，可以尝试更新deepspeed或更换使用的deepspeed配置文件

我换了一个虚拟环境：python3.7，cuda10.2，torch1.31.1，transformers4.30.2，deepspeed0.15.3，没有报上面的错了，但是还是OOM了，我按照你的指示设置了gradient_checkpoint为True，deepspeed的配置文件是
`{
    "fp16": {
        "enabled": "auto",
        "loss_scale": 0,
        "loss_scale_window": 1000,
        "initial_scale_power": 12,
        "hysteresis": 2,
        "min_loss_scale": 1
    },

    "bf16": {
        "enabled": "auto"
    },

    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "betas": "auto",
            "eps": "auto",
            "weight_decay": "auto"
        }
    },

    "scheduler": {
        "type": "WarmupDecayLR",
        "params": {
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto",
            "total_num_steps": "auto"
        }
    },

    "zero_optimization": {
        "stage": 0
    },

    "gradient**_ac**cumulation_steps": "auto",
    "gradient_clipping": "auto",
    "steps_per_print": 100,
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto",
    "wall_clock_breakdown": false
}`

`

**Comment by 545999961** [comment]
可以换个更高点的stage文件，如果还OOM的话可以考虑缩小batch size或negative数量等

**Comment by WangCC-77** [comment]
> 可以换个更高点的stage文件，如果还OOM的话可以考虑缩小batch size或negative数量等
@545999961 好的，我试试。目前环境只能使用stage=2的配置文件，然后我把batchsize设置得非常小了，文本长度(0-500)和(500-1000）是4，其余的是2。不过依然会报OOM，我接下来再把negative数量设置成5再试试，谢谢作者的耐心回复。

**Comment by WangCC-77** [comment]
> 可以换个更高点的stage文件，如果还OOM的话可以考虑缩小batch size或negative数量等

@545999961 作者，您好呀，我又来打扰你了，我上周又进行了实验：train_group=4，只有3个负样本了，batchsize=2，deepspeed的stage=2，gradient_checkpointing=True，gradient_accumulation_steps=4，dataloader_pin_memory=True，dataloader_num_workers=4，epoch=5，**在执行到36%时，差不多一个epoch的时候就报OOM了**，现在有点不知道往什么方向去尝试了，是要继续减小negative的数量吗，主要batchsize已经没办法再减小了，麻烦作者大大帮忙看看呢，感谢~

**Comment by xushan116** [comment]
@WangCC-77 同学你好，我现在也是想用V100的卡去微调bge-m3，目前也是报错，我想问下你跑通了吗，各种环境都是什么配置呀？python环境、cuda环境等等

**Comment by WangCC-77** [comment]
> 各种环境都是

@xushan116 
目前是可以跑通，但是跑了一个epoch后就报OOM了，不知道是不是缓存堆积的问题，想问问作者是不是有一些关于清理缓存的配置。
我的环境是python3.7，cuda10.2，torch1.31.1，transformers4.30.2，deepspeed0.15.3，然后deepspeed的配置文件用的stage=1的，然后还设置了一些减少显存的：fp16，gradient_checkpointing True ，gradient_accumulation_steps=4 ，dataloader_pin_memory=True ，dataloader_num_workers=4，然后negative的数量为3，其他的就没什么特别的了。有什么不清楚的你可以翻阅一下我前面和作者的交流哈，如果还有问题的话，你再给我留言吧

**Comment by xushan116** [comment]
@WangCC-77 torch是1.13.1吗

**Comment by xushan116** [comment]
@WangCC-77 我的现在去掉deep_speed 和gradient_checkpointing 这两设置，
用下面配置能跑通了 ，
V100卡
cuda10.1(nvcc -V显式的是这个)或者是cuda12.0(nvidia-smi显式的是12.0)
Pytho3.10
torch2.5.1
transformers4.44.2
deepspeed 0.15.4
你也可以试一下。

再请教一下，你知道train_group_size这个参数是什么含义吗？怎么设置？
还有就是需要用它给的脚本将文章按照不同长度划分吗，如果划分的话训练时参数设置有什么不一样吗？还是划分了就能直接用？

**Comment by WangCC-77** [comment]
> @WangCC-77 我的现在去掉deep_speed 和gradient_checkpointing 这两设置， 用下面配置能跑通了 ， V100卡 cuda10.1(nvcc -V显式的是这个)或者是cuda12.0(nvidia-smi显式的是12.0) Pytho3.10 torch2.5.1 transformers4.44.2 deepspeed 0.15.4 你也可以试一下。
> 
> 再请教一下，你知道train_group_size这个参数是什么含义吗？怎么设置？ 还有就是需要用它给的脚本将文章按照不同长度划分吗，如果划分的话训练时参数设置有什么不一样吗？还是划分了就能直接用？

（1）好的，谢谢，我下来试试。你跑通的是bge-m3这个微调模型吗，你的query和pos的长度分别是多少呀？我想要训练query为512，pos为8192的，目前跑模型出现OOM，估计就是长度太长了吧。
（2）train_group_size我看作者取的值是8，代表7个负样本，在代码里面写的是
![image](https://github.com/user-attachments/assets/1261afe7-997f-4498-be7c-de043786580b)
你可以再看看代码哈。
（3）关于第三个问题那个按长度划分我还没开始弄呢，我是直接用作者公开的数据集训练的，他给的数据集就是按长度划分好了
[https://huggingface.co/datasets/Shitao/bge-m3-data](url)

**Comment by xushan116** [comment]
@WangCC-77 对的，我跑通的是bge-m3这个模型

**Comment by LLLLucensus** [comment]
> > @WangCC-77 我的现在去掉deep_speed 和gradient_checkpointing 这两设置， 用下面配置能跑通了 ， V100卡 cuda10.1(nvcc -V显式的是这个)或者是cuda12.0(nvidia-smi显式的是12.0) Pytho3.10 torch2.5.1 transformers4.44.2 deepspeed 0.15.4 你也可以试一下。
> > 再请教一下，你知道train_group_size这个参数是什么含义吗？怎么设置？ 还有就是需要用它给的脚本将文章按照不同长度划分吗，如果划分的话训练时参数设置有什么不一样吗？还是划分了就能直接用？
> 
> （1）好的，谢谢，我下来试试。你跑通的是bge-m3这个微调模型吗，你的query和pos的长度分别是多少呀？我想要训练query为512，pos为8192的，目前跑模型出现OOM，估计就是长度太长了吧。 （2）train_group_size我看作者取的值是8，代表7个负样本，在代码里面写的是 ![image](https://private-user-images.githubusercontent.com/75521911/387904695-1261afe7-997f-4498-be7c-de043786580b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY5MjI5NjIsIm5iZiI6MTczNjkyMjY2MiwicGF0aCI6Ii83NTUyMTkxMS8zODc5MDQ2OTUtMTI2MWFmZTctOTk3Zi00NDk4LWJlN2MtZGUwNDM3ODY1ODBiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE1VDA2MzEwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYxMjcyYTg5Y2QxMDMxNzNjY2ZhMWJkYjk4YzI1M2E0NTNiMGUzZTBkMDE0NWVlYTQxYmZiMWQ3YjkxMmVjYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tb9apXD89oYVePsDyggKr5uzkCaTozpveZGBDcV9at4) 你可以再看看代码哈。 （3）关于第三个问题那个按长度划分我还没开始弄呢，我是直接用作者公开的数据集训练的，他给的数据集就是按长度划分好了 [https://huggingface.co/datasets/Shitao/bge-m3-data](url)

后来有加上这两配置deep_speed 和gradient_checkpointing吗，有跑通过吗，求版本配置
